File: fabric/src/main/java/mcpc/tedo0627/kzeaddon/fabric/CustomConfig.java
package mcpc.tedo0627.kzeaddon.fabric;

import com.google.gson.Gson;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonWriter;

import java.io.*;

public class CustomConfig {

    private static final File file = new File("config", "KZEAddon.json");

    public boolean deleteJoinMessage = false;
    public boolean deleteQuitMessage = false;

    public CustomConfig() {}

    public CustomConfig(boolean deleteJoinMessage, boolean deleteQuitMessage) {
        this.deleteJoinMessage = deleteJoinMessage;
        this.deleteQuitMessage = deleteQuitMessage;
    }

    public static CustomConfig load() {
        try {
            if (!file.exists()) {
                JsonWriter writer = new JsonWriter(new FileWriter(file));
                writer.setIndent("    ");
                new Gson().toJson(new CustomConfig(), CustomConfig.class, writer);
                writer.close();
            }
            JsonReader reader = new JsonReader(new FileReader(file));
            CustomConfig config = new Gson().fromJson(reader, CustomConfig.class);
            reader.close();
            return config;
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public boolean isDeleteJoinMessage() {
        return deleteJoinMessage;
    }

    public void setDeleteJoinMessage(boolean bool) {
        deleteJoinMessage = bool;
    }

    public boolean isDeleteQuitMessage() {
        return deleteQuitMessage;
    }

    public void setDeleteQuitMessage(boolean bool) {
        deleteQuitMessage = bool;
    }

    public void save() {
        try {
            JsonWriter writer = new JsonWriter(new FileWriter(file));
            writer.setIndent("    ");
            new Gson().toJson(this, CustomConfig.class, writer);
            writer.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: fabric/src/main/kotlin/mcpc/tedo0627/kzeaddon/fabric/CustomConfig.kt
package mcpc.tedo0627.kzeaddon.fabric

import com.google.gson.Gson
import com.google.gson.JsonObject
import com.google.gson.JsonPrimitive
import com.google.gson.stream.JsonWriter
import java.io.File
import java.io.FileReader
import java.io.FileWriter

object CustomConfig {

    private val json: JsonObject

    init {
        try {
            val file = File("config", "KZEAddon.json")
            if (!file.exists()) {
                val writer = JsonWriter(FileWriter(file))
                writer.setIndent("    ")
                Gson().toJson(JsonObject(), writer)
                writer.close()
            }

            val reader = FileReader(file)
            json = Gson().fromJson(reader, JsonObject::class.java)
            reader.close()
        } catch (e: Exception) {
            e.printStackTrace()
            throw e
        }
    }

    fun getString(key: String, default: String): String {
        return json.getAsJsonPrimitive(key)?.asString ?: default
    }

    fun getBoolean(key: String, default: Boolean): Boolean {
        return json.getAsJsonPrimitive(key)?.asBoolean ?: default
    }

    fun <T : Enum<T>> getEnum(key: String, default: Enum<T>): Enum<T> {
        val obj = json.getAsJsonPrimitive(key) ?: return default
        val ordinal = obj.asInt

        val enums = default::class.java.enumConstants
        if (enums.size <= ordinal) return default

        return enums[ordinal]
    }

    fun set(key: String, value: Boolean) {
        json.add(key, JsonPrimitive(value))
    }

    fun set(key: String, value: Number) {
        json.add(key, JsonPrimitive(value))
    }

    fun set(key: String, value: String) {
        json.add(key, JsonPrimitive(value))
    }

    fun set(key: String, value: Enum<*>) {
        json.add(key, JsonPrimitive(value.ordinal))
    }

    fun save() {
        try {
            val file = File("config", "KZEAddon.json")
            val writer = JsonWriter(FileWriter(file))
            writer.setIndent("    ")
            Gson().toJson(json, writer)
            writer.close()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
}